pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -589,7 +589,19 @@ func (c *Reconciler) runNextSchedulableTask(ctx context.Context, pr *v1beta1.Pip
 	pipelineRunFacts.ResetSkippedCache()
 
 	// GetFinalTasks only returns tasks when a DAG is complete
-	nextRprts = append(nextRprts, pipelineRunFacts.GetFinalTasks()...)
+	finallyRprts := pipelineRunFacts.GetFinalTasks()
+
+	// Before creating TaskRun for scheduled final task, check if it's consuming a task result
+	// Resolve and apply task result wherever applicable, report warning in case resolution fails
+	for _, rprt := range finallyRprts {
+		resolvedResultRefs, err := resources.ResolveResultRef(pipelineRunFacts.State, rprt)
+		if err != nil {
+			logger.Warnf("Final task %q is not executed as it failed to resolve task params for %q with error %v", rprt.PipelineTask.Name, pr.Name, err)
```

Comment:
should this be a warning or an error? we only need to apply results when we're ready to run final tasks and we're only ready once all of the non-final tasks have finished.

Reply:
im thinking that what is happening here is that we are logging this warning in the case where a finally task needs a result from a task that was skipped.

1. i dont think that should be a warning. logs that are warn, error or fatal should be indicating to the operator of the system that some kind of action needs to be taken. i.e. the _operator_ not the _user_. the user gets informed of actions they need to take via the status (e.g. error messages). the operator is the person admin-ing the tekton installation and needs to know if there is something wrong with the setup. so maybe info here? (https://dzone.com/articles/logging-levels-what-they-are-and-how-they-help-you seems like an okay article on log levels)
2. it would be great if one day we could get to a point where the decision of whether or not to skip was only done in one place - is there any way to do that in this PR? i.e. before we try to resolve the references? (i.e. why are we trying to resolve references on something that needs to be skipped) it seems like we should be able to look at a task's dependencies and see that it cant be run?

Reply:
> should this be a warning or an error?

I think it should be warning, its indicating that the finally task was skipped for some reason. Skipping a dag does not warn users but here users are warned when a finally task is skipped. Finally being finally task must be executed and warn if they are not for some reason.

But based on your description, there is no action needs to be taken if a finally task is skipped in case of a cleanup/notification use case. may be convert it to info then? I am not convinced but at the same time I dont see a use case.


> im thinking that what is happening here is that we are logging this warning in the case where a finally task needs a result from a task that was skipped.

yup skipped or failed.

> we only need to apply results when we're ready to run final tasks and we're only ready once all of the non-final tasks have finished.

Until we are ready to run finally tasks, `finallyRprts` is empty.

> it would be great if one day we could get to a point where the decision of whether or not to skip was only done in one place - is there any way to do that in this PR?

The only way we could do it to maintain a list of [skippedTasks](https://github.com/tektoncd/pipeline/compare/master...pritidesai:pipelinerunstatus#diff-188e8bbe2f5c7cb0b9de03be02a957537e46bb71add977e4992533375ef08998R351) in `pipelineRunFacts`. Once we determine that a finally task is being skipped, add it into the list. And check this list while updating the status condition and status.SkippedTasks. The same logic would work for dag tasks as well. We could merge this as is and refactor after this is merged.

(@afrittoli's attempt on simplifying [skipped](https://github.com/tektoncd/pipeline/commit/72455ba0a81c9d806b26809842282c045ca43d41#diff-2d6877ada24d6e8ae6d232d24799f21c520e590e5b432f9adba5e2e62eaa40b7R53) tasks)

> why are we trying to resolve references on something that needs to be skipped

At this point, we are discovering task result resolution failure and therefore not executing finally task. Finally task consuming task results from a dag task but not dependent on a dag task. Both belong to separate sections, no relationships established from a finally to dag.

Reply:
This could be further simplified for both finally and dag, instead of trying to resolve task results, just check if the referenced task was failed or skipped before trying to resolve results.